docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,585 @@
+* No spaces around arrow: `this->fnc();`.
+* No trailing whitespace.
+* Enums use `UPPER_CASE`.
+* Macros use `UPPER_CASE`.
+* Align attributes by group in classes, and initialisers.
+* Variables should be initialised with `=` if primitive, or braced initialisation with parentheses if required.
+* Initialiser lists should be initialised with braces by default.
+* Use SQL style initialiser for lists and inheritance lists.
+
+```c++
+Foo::Foo()
+  : attribute1_{value1}
+  , attribute2_{value2}
+{}
+```
+
+
+## Typing
+
+* Use STL types wherever possible.
+* Prefer `(u)int(16/32/64)_t` and `std::size_t`.
+* Use `static_cast`, `reinterpret_cast`, `dynamic_cast`, etc., rather than C-style casts.
+* Avoid `auto` except when the return type is clear.
+* Avoid use of RTTI.
+
+* `using` is preferable to `typedef` for type alias definition.
+* Make `using` private unless you intend to expose it.
+* `namespace {.<template T>` only allowed if `typedef`/`using` indicates the type.
+
+* If a value is `const`, declare it so.
+* Use `int const &i` over `const int &i`.
+
+
+## Namespaces 
+
+* Do not do `using namespace foo` or `namespace baz = ::foo::bar::baz` in `.hpp` files.
+* Do not do `using namespace foo` in the top `fetch` namespace, as it pollutes all of it. Instead, do the following:
+
+``` c++
+namespace fetch 
+{
+    namespace details 
+    {
+        struct FecthImpl { /* ... */ };
+      
+        void HelperFunction(FecthImpl & s) 
+        {
+           /* ... */
+        }  
+    };
+
+    class MyFetchClass {
+
+        public:
+
+            void DoSomeThing() {
+              // This improves readability
+              using namespace details; 
+
+              FetchImpl f;
+              HelperFunction(f);
+            }
+            // ...
+    };
+
+};
+```
+
+
+## Structs
+
+Use structs for Plain Old Data (POD) and, in some cases, template meta-programming. Use `classes` for everything else.
+
+``` c++
+struct Person 
+{
+    std::string name;
+    int age;
+    // ...
+};
+```
+
+
+## Classes
+
+* Initialise class members in the following order of preference:
+	1. Inline initialisation.
+	2. Constructor initialiser list.
+	3. Constructor body.
+
+* Mark all constructors and assignment operators explicitly, even those which are implicit, to ensure there is no confusion about the intended behaviour of the class.
+* Mark unneeded default implementations as deleted. C++ infers whether to create implicit constructors based on whether there are user defined ones. This means that when someone declares a constructor, C++ may remove the default constructor.
+
+``` c++
+class Transaction 
+{
+  public:
+      Transaction() = default;
+      Transaction(Transaction const&) = default;
+      Transaction(Transaction &&) = deleted;
+
+      ~Transaction() = default;
+
+      Transaction operator(Transaction const&) = default;
+      Transaction operator(Transaction &&) = deleted;
+}
```

Comment:
Identation - specifically we don't align the `public` /  `private` keywords either